Source/Core/Proj4Projection.js

```diff
@@ -0,0 +1,196 @@
+define([
+        './Cartesian3',
+        './Cartographic',
+        './Math',
+        './Check',
+        './defaultValue',
+        './defined',
+        './defineProperties',
+        './Ellipsoid',
+        './oneTimeWarning',
+        './Rectangle',
+        '../ThirdParty/proj4-2.5.0'
+    ], function(
+        Cartesian3,
+        Cartographic,
+        CesiumMath,
+        Check,
+        defaultValue,
+        defined,
+        defineProperties,
+        Ellipsoid,
+        oneTimeWarning,
+        Rectangle,
+        proj4) {
+    'use strict';
+
+    // proj4 may crash with absolute extents
+    var smallerThanMax = Rectangle.fromDegrees(-180.0 + CesiumMath.EPSILON7, -90.0 + CesiumMath.EPSILON7, 180 - CesiumMath.EPSILON7, 90 - CesiumMath.EPSILON7);
+
+    /**
+     * MapProjection using proj4js. This projection is only to be used with Ellipsoid.WGS84.
+     * Users should exercise caution when using local-area projections, as local area projections
+     * may produce unexpected results outside their valid boundaries in Cartographic space (wgs84Bounds).
+     * These boundaries may be looked up at spatialreference.org.
+     *
+     * Scenes using Proj4Projection will default to <code>MapMode2D.ROTATE</code> instead of <code>MapMode2D.INFINITE_SCROLL</code>.
+     *
+     * @alias Proj4Projection
+     * @constructor
+     *
+     * @param {String} [wellKnownText] proj4js well known text specifying the projection. Defaults to EPSG:3857, web mercator.
+     * @param {Number} [heightScale=1.0] Scale to convert from heights in meters to the projection's units.
+     * @param {Rectangle} [wgs84Bounds] Cartographic bounds over which the projection is valid. Cartographic points will be clamped to these bounds prior to projection.
+     */
+    function Proj4Projection(wellKnownText, heightScale, wgs84Bounds) {
+        this.ellipsoid = Ellipsoid.WGS84;
+
+        var wkt = defaultValue(wellKnownText, 'EPSG:3857'); // web mercator
+
+        this._projection = proj4(wkt);
+        this._wkt = wkt;
+
+        heightScale = defaultValue(heightScale, 1.0);
+        this._heightScale = heightScale;
+        this._inverseHeightScale = 1.0 / heightScale;
+
+        wgs84Bounds = defaultValue(wgs84Bounds, smallerThanMax);
```

Comment:
From a code style standpoint, Cesium usually sets all the defaults at the top of the function and all of the `this` assignments at the bottom in one go (it's easier to read than threading them through like this.